File: samples/src/main/java/io/github/danielbunting/clickhouse/samples/QuickStart.java
package io.github.danielbunting.clickhouse.samples;

import io.github.danielbunting.clickhouse.BulkInserter;
import io.github.danielbunting.clickhouse.ClickHouseConnection;
import io.github.danielbunting.clickhouse.QueryResult;
import io.github.danielbunting.clickhouse.protocol.Block;

import java.util.List;

/**
 * Minimal end-to-end demo of the clickhouse-native-client library.
 *
 * <p>Steps performed:
 * <ol>
 *   <li>Connect to ClickHouse using {@link ClickHouseEnv}.</li>
 *   <li>Create (or reuse) the table {@code events}.</li>
 *   <li>Bulk-insert five sample rows via {@link BulkInserter}.</li>
 *   <li>Query the rows back and print each one.</li>
 *   <li>Print the row count returned by {@code SELECT count()}.</li>
 * </ol>
 *
 * <p>Run with:
 * <pre>{@code
 *   ./gradlew :samples:runQuickStart
 * }</pre>
 *
 * <p>Override the target server with environment variables:
 * {@code CH_HOST}, {@code CH_PORT}, {@code CH_DB}, {@code CH_USER},
 * {@code CH_PASSWORD}.
 */
public final class QuickStart {

    /**
     * A record whose component names match the columns of the {@code events} table.
     * The {@link BulkInserter} mapper resolves columns by component name.
     *
     * @param id    row identifier
     * @param name  human-readable label
     * @param score arbitrary floating-point metric
     */
    public record Event(long id, String name, double score) {}

    /** The table targeted by this demo. */
    private static final String TABLE = "events";

    /** Private constructor — utility-style entry point. */
    private QuickStart() {}

    /**
     * Entry point. Connects, writes, reads, and prints results.
     *
     * @param args ignored
     */
    public static void main(String[] args) {
        System.out.println("=== clickhouse-native-client QuickStart ===");
        System.out.println("Connecting to ClickHouse at "
                + ClickHouseEnv.config().host() + ":" + ClickHouseEnv.config().port()
                + " (db=" + ClickHouseEnv.config().database() + ")");

        try (ClickHouseConnection conn = ClickHouseEnv.open()) {

            // ----------------------------------------------------------------
            // 1. DDL: ensure the table exists
            // ----------------------------------------------------------------
            System.out.println("\n[1] Creating table '" + TABLE + "' if not exists...");
            conn.execute(
                    "CREATE TABLE IF NOT EXISTS " + TABLE + " ("
                    + "id    UInt64,"
                    + "name  String,"
                    + "score Float64"
                    + ") ENGINE = MergeTree ORDER BY id");
            System.out.println("    Done.");

            // ----------------------------------------------------------------
            // 2. Bulk insert sample rows
            // ----------------------------------------------------------------
            List<Event> rows = List.of(
                    new Event(1L, "alpha",   9.5),
                    new Event(2L, "beta",    7.3),
                    new Event(3L, "gamma",   8.1),
                    new Event(4L, "delta",   6.0),
                    new Event(5L, "epsilon", 10.0)
            );

            System.out.println("\n[2] Bulk-inserting " + rows.size() + " rows...");
            try (BulkInserter<Event> inserter = conn.createBulkInserter(TABLE, Event.class)) {
                inserter.init();
                inserter.addRange(rows);
                inserter.complete();
            }
            System.out.println("    Inserted " + rows.size() + " rows.");

            // ----------------------------------------------------------------
            // 3. Query back and print each row
            // ----------------------------------------------------------------
            System.out.println("\n[3] Querying rows (SELECT id, name, score FROM "
                    + TABLE + " ORDER BY id):");
            System.out.printf("    %-6s %-12s %s%n", "id", "name", "score");
            System.out.println("    " + "-".repeat(30));

            try (QueryResult result = conn.query(
                    "SELECT id, name, score FROM " + TABLE + " ORDER BY id")) {

                var blockIter = result.blocks();
                while (blockIter.hasNext()) {
                    Block block = blockIter.next();
                    if (block.isEmpty()) {
                        continue;
                    }
                    for (int row = 0; row < block.rowCount(); row++) {
                        Object idVal    = block.column(0).value(row);
                        Object nameVal  = block.column(1).value(row);
                        Object scoreVal = block.column(2).value(row);
                        System.out.printf("    %-6s %-12s %s%n", idVal, nameVal, scoreVal);
                    }
                }
            }

            // ----------------------------------------------------------------
            // 4. Scalar count
            // ----------------------------------------------------------------
            long count = conn.executeScalar("SELECT count() FROM " + TABLE);
            System.out.println("\n[4] Total rows in '" + TABLE + "': " + count);

        }

        System.out.println("\n=== QuickStart complete ===");
    }
}

File: samples/src/main/kotlin/io/github/danielbunting/clickhouse/samples/kotlin/QuickStart.kt
package io.github.danielbunting.clickhouse.samples.kotlin

import io.github.danielbunting.clickhouse.samples.ClickHouseEnv
import io.github.danielbunting.clickhouse.kotlin.bulkInsert
import io.github.danielbunting.clickhouse.kotlin.command
import io.github.danielbunting.clickhouse.kotlin.connect
import io.github.danielbunting.clickhouse.kotlin.queryAs
import io.github.danielbunting.clickhouse.kotlin.scalar
import kotlinx.coroutines.runBlocking

/**
 * Minimal end-to-end demo of the Kotlin coroutine surface — the counterpart of the Java
 * [QuickStart] sample.
 *
 * Steps:
 *  1. Connect (suspend) using the shared [ClickHouseEnv] settings.
 *  2. Create (or reuse) the `events_kotlin` table via `command`.
 *  3. Bulk-insert five rows of an **immutable** data class.
 *  4. Stream the rows back with `queryAs` (primary-constructor binding) and print them.
 *  5. Print `count()` via `scalar`.
 *
 * Run with:
 * ```
 *   ./gradlew :samples:runKotlinQuickStart
 * ```
 * Override the target server with `CH_HOST`, `CH_PORT`, `CH_DB`, `CH_USER`, `CH_PASSWORD`.
 */

/** Component names match the table's columns; `val`s only — no no-arg constructor needed. */
data class Event(val id: Long, val name: String, val score: Double)

fun main(): Unit = runBlocking {
    val table = "events_kotlin"
    println("=== clickhouse-native-client-kotlin QuickStart ===")
    val config = ClickHouseEnv.config()
    println("Connecting to ${config.host()}:${config.port()} (db=${config.database()})")

    config.connect().use { conn ->

        println("\n[1] Creating table '$table' if not exists...")
        conn.command(
            "CREATE TABLE IF NOT EXISTS $table (" +
                "id    UInt64," +
                "name  String," +
                "score Float64" +
                ") ENGINE = MergeTree ORDER BY id",
        )
        println("    Done.")

        val rows = listOf(
            Event(1, "alpha", 9.5),
            Event(2, "beta", 7.3),
            Event(3, "gamma", 8.1),
            Event(4, "delta", 6.0),
            Event(5, "epsilon", 10.0),
        )
        println("\n[2] Bulk-inserting ${rows.size} rows...")
        conn.bulkInsert(table, rows)
        println("    Inserted ${rows.size} rows.")

        println("\n[3] Querying rows (queryAs<Event>: SELECT id, name, score FROM $table ORDER BY id):")
        println("    %-6s %-12s %s".format("id", "name", "score"))
        println("    " + "-".repeat(30))
        conn.queryAs<Event>("SELECT id, name, score FROM $table ORDER BY id")
            .collect { println("    %-6d %-12s %s".format(it.id, it.name, it.score)) }

        println("\n[4] Total rows in '$table': ${conn.scalar("SELECT count() FROM $table")}")
    }

    println("\n=== Kotlin QuickStart complete ===")
}
